File: dotnet/src/DocGen/DocGenConfig.cs
namespace Dokmatiq.DocGen;

/// <summary>
/// Configuration for the DocGen client.
/// </summary>
public sealed class DocGenConfig
{
    /// <summary>API key (required). Starts with "dk_live_" or "dk_test_".</summary>
    public string ApiKey { get; }

    /// <summary>Base URL of the DocGen API. Default: https://api.dokmatiq.com</summary>
    public string BaseUrl { get; init; } = "https://api.dokmatiq.com";

    /// <summary>Request timeout. Default: 60 seconds.</summary>
    public TimeSpan Timeout { get; init; } = TimeSpan.FromSeconds(60);

    /// <summary>Retry policy for transient failures.</summary>
    public RetryPolicy Retry { get; init; } = RetryPolicy.Default;

    /// <summary>Validation mode: "strict" or "warn". Null for default behavior.</summary>
    public string? ValidateMode { get; init; }

    public DocGenConfig(string apiKey)
    {
        ApiKey = apiKey ?? throw new ArgumentNullException(nameof(apiKey));
    }
}

/// <summary>
/// Retry policy for transient HTTP errors (429, 500, 502, 503, 504).
/// </summary>
public sealed class RetryPolicy
{
    /// <summary>Maximum number of retries. Default: 3.</summary>
    public int MaxRetries { get; init; } = 3;

    /// <summary>Initial delay before first retry. Default: 500ms.</summary>
    public TimeSpan InitialDelay { get; init; } = TimeSpan.FromMilliseconds(500);

    /// <summary>Backoff multiplier. Default: 2.0.</summary>
    public double BackoffMultiplier { get; init; } = 2.0;

    /// <summary>Maximum delay between retries. Default: 30 seconds.</summary>
    public TimeSpan MaxDelay { get; init; } = TimeSpan.FromSeconds(30);

    /// <summary>Default retry policy.</summary>
    public static RetryPolicy Default => new();
}

File: java/src/main/java/com/dokmatiq/docgen/DocGenConfig.java
package com.dokmatiq.docgen;

import java.time.Duration;
import java.util.Objects;

/** DocGen client configuration. */
public final class DocGenConfig {
    private final String apiKey;
    private final String baseUrl;
    private final Duration timeout;
    private final RetryPolicy retry;
    private final String validateMode;

    private DocGenConfig(Builder builder) {
        this.apiKey = Objects.requireNonNull(builder.apiKey, "apiKey is required");
        this.baseUrl = builder.baseUrl != null ? builder.baseUrl : "https://api.dokmatiq.com";
        this.timeout = builder.timeout != null ? builder.timeout : Duration.ofSeconds(60);
        this.retry = builder.retry != null ? builder.retry : RetryPolicy.defaultPolicy();
        this.validateMode = builder.validateMode;
    }

    public static Builder builder(String apiKey) { return new Builder(apiKey); }

    public String apiKey() { return apiKey; }
    public String baseUrl() { return baseUrl; }
    public Duration timeout() { return timeout; }
    public RetryPolicy retry() { return retry; }
    public String validateMode() { return validateMode; }

    /** Retry policy configuration. */
    public record RetryPolicy(
            int maxRetries,
            Duration initialDelay,
            double backoffMultiplier,
            Duration maxDelay
    ) {
        public static RetryPolicy defaultPolicy() {
            return new RetryPolicy(3, Duration.ofMillis(500), 2.0, Duration.ofSeconds(30));
        }
    }

    public static class Builder {
        private final String apiKey;
        private String baseUrl;
        private Duration timeout;
        private RetryPolicy retry;
        private String validateMode;

        Builder(String apiKey) { this.apiKey = apiKey; }

        public Builder baseUrl(String baseUrl) { this.baseUrl = baseUrl; return this; }
        public Builder timeout(Duration timeout) { this.timeout = timeout; return this; }
        public Builder retry(RetryPolicy retry) { this.retry = retry; return this; }
        public Builder validateMode(String mode) { this.validateMode = mode; return this; }

        public DocGenConfig build() { return new DocGenConfig(this); }
    }
}
